readme.md
# Advanced ImagePicker Cordova Plugin
![Maintenance](https://img.shields.io/maintenance/yes/2022)
[![npm version](https://badge.fury.io/js/cordova-plugin-advanced-imagepicker.svg)](https://badge.fury.io/js/cordova-plugin-advanced-imagepicker)

This [Cordova](https://cordova.apache.org) Plugin is for a better (multiple) ImagePicker with more options.

It currently uses [Yummypets/YPImagePicker](https://github.com/Yummypets/YPImagePicker) (Version `5.2.1`) on iOS and 
[ParkSangGwon/TedImagePicker](https://github.com/ParkSangGwon/TedImagePicker) (Default-Version `1.2.8`) on Android. 

**This Plugin is in active development!**

<!-- DONATE -->
[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG_global.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=LMX5TSQVMNMU6&source=url)

This and other Open-Source Cordova Plugins are developed in my free time.
To help ensure this plugin is kept updated, new features are added and bugfixes are implemented quickly, please donate a couple of dollars (or a little more if you can stretch) as this will help me to afford to dedicate time to its maintenance.
Please consider donating if you're using this plugin in an app that makes you money, if you're being paid to make the app, if you're asking for new features or priority bug fixes.
<!-- END DONATE -->

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Content**

- [Install](#install)
  - [Requirements](#requirements)
  - [Android](#android)
  - [iOS](#ios)
- [Environment Variables](#environment-variables)
  - [Android](#android-1)
  - [iOS](#ios-1)
- [Usage](#usage)
  - [Failure Callbacks](#failure-callbacks)
  - [Error Codes](#error-codes)
- [Api](#api)
  - [All platforms](#all-platforms)
    - [present](#present)
  - [iOS](#ios-2)
    - [cleanup](#cleanup)
- [Quirks](#quirks)
  - [Android](#android-2)
- [Changelog](#changelog)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# Install

## Requirements

- **cordova** `>= 9.0.0`
- **cordova-android** `>= 9.0.0`

## Android
Because the used Framework uses AndroidX and is developed in Kotlin, make sure to enable `AndroidXEnabled` and `GradlePluginKotlinEnabled`: 
```xml
<preference name="AndroidXEnabled" value="true"/>
<preference name="GradlePluginKotlinEnabled" value="true"/>
```

## iOS

This Plugin is developed in Swift and automaticaly adds the Plugin to [Support Swift](https://github.com/akofman/cordova-plugin-add-swift-support).

I developed it, testing with **cordova-ios@6.1.0**.

# Environment Variables

## Android

- ANDROID_IMAGE_PICKER_VERSION - Version of `gun0912.ted:tedimagepicker` / default to `1.2.8` 

## iOS

The iOS platform defines:
- NSCameraUsageDescription: **This app requires access to the camera to take pictures and videos.**
- NSPhotoLibraryUsageDescription: **This app requires access to the photo library to select pictures and videos.**
- NSMicrophoneUsageDescription: **This app requires access to the microphone to record videos.**

You can easily change them, by configure your **config.xml** by:
```xml
<edit-config file="*-Info.plist" mode="merge" target="NSCameraUsageDescription">
    <string>your text</string>
</edit-config>
```

# Usage

The plugin is available via a global variable named `window.AdvancedImagePicker`.
A TypeScript definition is included out of the Box. You can import it like this:
```ts
import AdvancedImagePicker from 'cordova-plugin-advanced-imagepicker';
```

## Failure Callbacks

If an Error appeared this Plugin returns an Object in the failureCallback, that always has the following Structure:

```json
{
  "code": 0,
  "message": "Some additional Info"
}
```

The `code` is one of the [Error Codes](#error-codes) and always present, while the `message` can be empty.
This is mostly something like an Exception Message.

## Error Codes

The following Error Codes can be fired by this Plugin:
- UnsupportedAction
- WrongJsonObject
- PickerCanceled  
- UnknownError

They can be accessed over for Example `window.AdvancedImagePicker.ErrorCodes.UnsupportedAction` and are present in the TypeScript definition too of course. 

# Api

The list of available methods for this plugin is described below.

## All platforms

### present

Open the ImagePicker

#### Parameters:

- options (object) - a JSON-Object containing the following Elements (all optional):
    - mediaType (string) **default: "IMAGE"**
    - showCameraTile (boolean) **default: true**
    - startOnScreen (string) **default: "LIBRARY"**
    - scrollIndicatorDateFormat (string) **Android only, default: "YYYY.MM"**
    - showTitle (boolean) **Android only, default: true**
    - title (string) **Android only, default: "Select Image"**
    - zoomIndicator (boolean) **Android only, default: true**
    - min (number) **default: 0 (android), 1 (iOS)**
    - minCountMessage (string) **Android only, default: "You need to select a minimum of ... files"**
    - max (number) **default: 0 (android), 1 (iOS)**
    - maxCountMessage (string) **default: "You need to select a minimum of ... files"**
    - buttonText (string)
    - asDropdown (boolean) **default: false**
    - asBase64 (boolean) **default: false**
    - asJpeg (boolean) **default: false**
    - videoCompression (string) ([Available Options](https://github.com/Yummypets/YPImagePicker/blob/23158e138bd649b40762bf2e4aa4beb0d463a121/Source/Configuration/YPImagePickerConfiguration.swift#L226-L240)) **default: AVAssetExportPresetHighestQuality**
    - recordingTimeLimit (number) **default: 60.0** **IOS only**
    - libraryTimeLimit (number) **default: 60.0** **IOS only**
    - minimumTimeLimit (number) **default: 3.0** **IOS only**

```js
window.AdvancedImagePicker.present({
  // config here
}, function(success) {
  console.log(success);
}, function (error) {
  console.error(error);
});
```
#### SuccessType:

This Methode returns an Array of Objects with the following fields:

- isBase64 (boolean) - is file base64 encoded
- type (`'image'` | `'video'`) - type of selected file
- src (string) - Result as file-uri or base64 encoded string

## iOS

### cleanup

Cleans the files that were stored in the tmp file directory

```js
window.AdvancedImagePicker.cleanup(function() {
  console.log('success');
}, function (error) {
  console.error(error);
});
```

# Quirks

## Android

- Currently, the Android Library is not able to select Images and Videos at the same Time. See reported Issue [here](https://github.com/ParkSangGwon/TedImagePicker/issues/40).

# Changelog

The full Changelog is available [here](CHANGELOG.md)


CHANGELOG.md
### [1.6.2](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/compare/V1.6.1...V1.6.2) (2022-08-16)


### Bug Fixes

* commit `plugin.xml` automatically ([ad0186a](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/commit/ad0186afb6f9d62d98656b12c527910efe1c2790))

### [1.6.1](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/compare/V1.6.0...V1.6.1) (2022-08-16)


### Code Refactoring

* correctly set plugin version in `plugin.xml` ([397b515](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/commit/397b515280d4bf90a053e892ec41eaa3c292ee74))

## [1.6.0](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/compare/V1.5.5...V1.6.0) (2022-08-16)


### ⚠ BREAKING CHANGES

* **ios:** you have to manually set `NSCameraUsageDescription`, `NSPhotoLibraryUsageDescription` and `NSMicrophoneUsageDescription` now

### Features

* **android:** Support PickerCanceled error ([#43](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/issues/43)) ([f57f470](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/commit/f57f47061c92edfd94ec37cbcb9300c941e55690))
* **ios:** update library version ([ca3a33e](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/commit/ca3a33e88e43d249b6aea57aef24136850f80415))


### Code Refactoring

* **ios:** remove hardcoded plist permission strings ([12425c7](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/commit/12425c74597e1b7d0a359c4c01d8942ba0f898fa))

### [1.5.5](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/compare/V1.5.4...V1.5.5) (2022-01-07)


### Bug Fixes

* **android:** don't wrap base64 output ([#33](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/issues/33)) ([b03620b](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/commit/b03620b429551c68c939ef484e692a665f3b29ca))

### [1.5.4](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/compare/V1.5.3...V1.5.4) (2021-10-18)


### Bug Fixes

* **ios:** use correct key for error code ([57180f3](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/commit/57180f356ee2034feedad22a6f0eca21eadd34b1))

### [1.5.3](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/compare/V1.5.2...V1.5.3) (2021-10-18)


### Bug Fixes

* **ios:** use double instead of NSFloat ([8bf948f](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/commit/8bf948f4af4ed6c686921eb4b1430d403ee2d4a4))

### [1.5.2](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/compare/V1.5.1...V1.5.2) (2021-10-18)


### Bug Fixes

* **ios:** fix ioa15 styling [See](https://github.com/Yummypets/YPImagePicker/issues/690) ([1c02839](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/commit/1c02839d2e6d9b46a61a532d056f67a5bf2f9561))
* **ios:** remove line breaks from generated base64 ([dfb472b](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/commit/dfb472b327696302b8e730b1832b93b78966bcf6))
* **ios:** remove line breaks from generated video base64 ([1f60c2d](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/commit/1f60c2ddec3051125c69aa9b3ec3b1fe8b00f081))

### [1.5.1](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/compare/V1.5.0...V1.5.1) (2021-10-12)


### Bug Fixes

* **ios:** use double instead of NSFloat ([#26](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/issues/26)) ([91a0c2a](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/commit/91a0c2af8f7bf38ebec6148720633186e921482d))

## [1.5.0](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/compare/V1.4.0...V1.5.0) (2021-10-10)


### Features

* **ios:** add more video config options ([#24](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/issues/24)) ([2833f38](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/commit/2833f38b35c63afe6eabe5b80ae53c849e4b072d))


### Code Refactoring

* add semantic release ([cdf0006](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/commit/cdf0006aea589d9444716b6184f60b8816689ac3))

## 1.4.0
- Upgrade iOS dependency ([See changes](https://github.com/Yummypets/YPImagePicker/releases/tag/4.5.0))
- Upgrade default android dependency ([See Changes](https://github.com/ParkSangGwon/TedImagePicker/releases/tag/1.2.2))
- iOS: `PickerCanceled` ErrorCode added

### Breaking
The Change of the android dependency also includes a dependency-name change. This **could** cause old versions to not work anymore. So make sure to update `ANDROID_IMAGE_PICKER_VERSION` in the `package.json` to `1.2.2`

## 1.3.2
- Change obsolete `android.dataBinding.enabled` to `android.buildFeatures.dataBinding`

## 1.3.1
- fix error with new android default version

## 1.3.0
- use image.pngData() instead of UIImagePNGRepresentation
- iOS: Video Compression added
- `asJpeg added`
- correctly compress Videos on Android

## 1.2.0
- return Unsupported Error on android
- fix: append correct screens by mediaType on iOS [#7](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/issues/7)
- `startOnScreen` config added for iOS

## 1.1.0
- Update Android default Version to 1.1.4 --> Enabled SDK 30 support
- Update iOS Pod to 4.4.0 --> [Changelog](https://github.com/Yummypets/YPImagePicker/releases)
- **BREAKING**: iOS know works with a file url and only returns isBase64 if `asBase64: true` was set
- Function `cleanup` added for ios


## 1.0.1
- I changed my GitHub Username, so i need to adjust stuff
- Typing fixed (closes [#4](https://github.com/EinfachHans/cordova-plugin-advanced-imagepicker/issues/4))

## 1.0.0
- Initial Release


src/ios/AdvancedImagePicker.swift
import YPImagePicker

@objc(AdvancedImagePicker) class AdvancedImagePicker : CDVPlugin  {

    var _callbackId: String?
    var OWN_PREFIX: String?

    @objc(pluginInitialize)
    override func pluginInitialize() {
        super.pluginInitialize()
        self.OWN_PREFIX = "advanced_image_picker_";
    }

    @objc(present:)
    func present(command: CDVInvokedUrlCommand) {
        _callbackId = command.callbackId;
        let options = command.argument(at: 0) as? NSDictionary;
        if(options == nil) {
            self.returnError(error: ErrorCodes.WrongJsonObject, message: "The first Argument must be the Configuration");
            return;
        }

        let mediaType = options?.value(forKey: "mediaType") as? String ?? "IMAGE";
        let startOnScreen = options?.value(forKey: "startOnScreen") as? String ?? "LIBRARY";
        let showCameraTile = options?.value(forKey: "showCameraTile") as? Bool ?? true;
        let min = options?.value(forKey: "min") as? NSInteger ?? 1;
        let max = options?.value(forKey: "max") as? NSInteger ?? 1;
        let defaultMaxCountMessage = "You can select a maximum of " + String(max) + " files";
        let maxCountMessage = options?.value(forKey: "maxCountMessage") as? String ?? defaultMaxCountMessage;
        let buttonText = options?.value(forKey: "buttonText") as? String ?? "";
        let asBase64 = options?.value(forKey: "asBase64") as? Bool ?? false;
        let videoCompression = options?.value(forKey: "videoCompression") as? String ?? "AVAssetExportPresetHighestQuality";
        let asJpeg = options?.value(forKey: "asJpeg") as? Bool ?? false;
        let recordingTimeLimit = options?.value(forKey: "recordingTimeLimit") as? Double ?? 60.0;
        let libraryTimeLimit = options?.value(forKey: "libraryTimeLimit") as? Double ?? 60.0;
        let minimumTimeLimit = options?.value(forKey: "minimumTimeLimit") as? Double ?? 3.0;

        if(max < 0 || min < 0) {
            self.returnError(error: ErrorCodes.WrongJsonObject, message: "Min and Max can not be less then zero.");
            return;
        }

        if(max < min) {
            self.returnError(error: ErrorCodes.WrongJsonObject, message: "Max can not be smaller than Min.");
            return;
        }

        var config = YPImagePickerConfiguration();
        config.onlySquareImagesFromCamera = false;
        config.showsPhotoFilters = false;
        config.showsVideoTrimmer = false;
        config.shouldSaveNewPicturesToAlbum = false;
        config.albumName = Bundle.main.infoDictionary![kCFBundleNameKey as String] as! String;
        config.library.isSquareByDefault = false;
        config.library.itemOverlayType = .none;
        config.library.skipSelectionsGallery = true;
        config.library.preSelectItemOnMultipleSelection = false;
        config.video.compression = videoCompression;
        config.video.recordingTimeLimit = recordingTimeLimit;
        config.video.libraryTimeLimit = libraryTimeLimit;
        config.video.minimumTimeLimit = minimumTimeLimit;

        if(startOnScreen == "IMAGE") {
            config.startOnScreen = .photo;
        } else if(startOnScreen == "VIDEO") {
            config.startOnScreen = .video;
        } else {
            config.startOnScreen = .library;
        }

        var screens: [YPPickerScreen] = [.library];
        if(showCameraTile) {
            if(mediaType != "VIDEO") {
                screens.append(.photo);
            }
            if(mediaType != "IMAGE") {
                screens.append(.video);
            }
        }
        config.screens = screens;
        config.library.defaultMultipleSelection = (max > 1);
        if(mediaType == "IMAGE") {
            config.library.mediaType = YPlibraryMediaType.photo
        } else if(mediaType == "VIDEO") {
            config.library.mediaType = YPlibraryMediaType.video
        } else {
            config.library.mediaType = YPlibraryMediaType.photoAndVideo
        }
        config.library.minNumberOfItems = min;
        config.library.maxNumberOfItems = max;
        config.wordings.warningMaxItemsLimit = maxCountMessage;
        if(buttonText != "") {
            config.wordings.next = buttonText;
        }

        let picker = YPImagePicker(configuration: config);

        if #available(iOS 15.0, *) {
            let navBarAppearance = UINavigationBarAppearance()
            navBarAppearance.configureWithOpaqueBackground()
            picker.navigationBar.scrollEdgeAppearance = navBarAppearance
        }

        picker.didFinishPicking {items, cancelled in
            if(cancelled) {
                self.returnError(error: ErrorCodes.PickerCanceled)
            } else if(items.count > 0) {
                self.handleResult(items: items, asBase64: asBase64, asJpeg: asJpeg);
            }
            picker.dismiss(animated: true, completion: nil);
        }

        self.viewController.present(picker, animated: true, completion: nil);
    }

    func handleResult(items: [YPMediaItem], asBase64: Bool, asJpeg: Bool) {
        var array = [] as Array;
        for item in items {
            switch item {
            case .photo(let photo):
                let encodedImage = self.encodeImage(image: photo.image, asBase64: asBase64, asJpeg: asJpeg);
                array.append([
                    "type": "image",
                    "isBase64": asBase64,
                    "src": encodedImage
                ]);
                break;
            case .video(let video):
                var resultSrc:String;
                if(asBase64) {
                    resultSrc = self.encodeVideo(url: video.url);
                    if(resultSrc == "") {
                        self.returnError(error: ErrorCodes.UnknownError, message: "Failed to encode Video")
                        return;
                    }
                } else {
                    resultSrc = video.url.absoluteString;
                }
                array.append([
                    "type": "video",
                    "isBase64": asBase64,
                    "src": resultSrc
                ]);
                break;
            }
        }
        let result:CDVPluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: array);
        self.commandDelegate.send(result, callbackId: _callbackId)
    }

    func encodeImage(image: UIImage, asBase64: Bool, asJpeg: Bool) -> String {
        let imageData: NSData;
        if(asJpeg) {
            imageData = UIImageJPEGRepresentation(image, 0.8)! as NSData;
        } else {
            imageData = UIImagePNGRepresentation(image)! as NSData;
        }
        if(asBase64) {
            return imageData.base64EncodedString();
        } else {
            let filePath = self.tempFilePath();
            do {
                try imageData.write(to: filePath, options: .atomic);
                return filePath.absoluteString;
            } catch {
                return error.localizedDescription;
            }
        }
    }

    func tempFilePath(ext: String = "png") -> URL {
        let filename: String = self.OWN_PREFIX! + UUID().uuidString;
        var contentUrl = URL(fileURLWithPath: NSTemporaryDirectory());
        contentUrl.appendPathComponent(filename);
        contentUrl.appendPathExtension(ext);
        return contentUrl;
    }

    func encodeVideo(url: URL) -> String {
        do {
            let fileData = try Data.init(contentsOf: url)
            return fileData.base64EncodedString();
        } catch {
            return "";
        }
    }

    func returnError(callbackId: String?, error: ErrorCodes, message: String = "") {
        if(callbackId != nil) {
            let result:CDVPluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR, messageAs: [
                    "code": error.rawValue,
                    "message": message
            ]);
            self.commandDelegate.send(result, callbackId: callbackId)
        }
    }

    func returnError(error: ErrorCodes, message: String = "") {
        self.returnError(callbackId: _callbackId, error: error, message: message)
        _callbackId = nil;
    }

    @objc(cleanup:)
    func cleanup(command: CDVInvokedUrlCommand) {
        do {
            let tmpFiles: [String] = try FileManager().contentsOfDirectory(atPath: NSTemporaryDirectory());
            for tmpFile in tmpFiles {
                // only delete files from this plugin:
                if(tmpFile.hasPrefix(self.OWN_PREFIX!)) {
                    try FileManager().removeItem(atPath: NSTemporaryDirectory() + tmpFile)
                }
            }
        } catch {
            returnError(callbackId: command.callbackId, error: ErrorCodes.UnknownError, message: error.localizedDescription);
            return;
        }
        let result:CDVPluginResult = CDVPluginResult(status: CDVCommandStatus_OK);
        self.commandDelegate.send(result, callbackId: command.callbackId);
    }
}

enum ErrorCodes:NSNumber {
    case UnsupportedAction = 1
    case WrongJsonObject = 2
    case PickerCanceled = 3
    case UnknownError = 10
}
